### ISSUE TITLE
Help Needed Exp Replication

### ISSUE THREAD

**Issue by anasserhussien** [yelboudouri/EmoNeXt#1]
Hello,

Thanks for providing the codebase.

I was trying to replicate the experiment with the "tiny" model. I followed the steps provided for downloading the dataset and training the model. However, the training always stops at epoch 12 because the validation loss didn't improve. Any idea/guidance?

Best,
Ahmed

**Comment by yelboudouri** [comment]
Hi Ahmed, thanks for your interest in using our model.

The problem you're facing with the training stopping at epoch 12 is a common occurrence that we've encountered as well during the development of EmoNeXt. To address this, you can adjust the early_stopping_patience parameter in the [Trainer](https://github.com/yelboudouri/EmoNeXt/blob/master/train.py#L39) class. By default, it's set to 12, but you can manually change it after cloning the code. Increasing the patience should allow the training to continue even if the validation loss doesn't improve immediately.

Please note that training the "tiny" model requires a considerable amount of time, at least 8 hours on a single GPU. So, you will have to be patient.

We believe that this behavior may be influenced by how the model weights are initialized. As an additional suggestion, you might want to experiment with a larger batch size during training, as this could potentially contribute to more stable convergence.

**Comment by anasserhussien** [comment]
Hi Yassine,

Thanks for the quick response!
I thought since the experiment is using the same seed as you assigned and the training is set to be deterministic I should get the same results.

But anyway I am trying to manipulate the `early_stopping_patience` and the `batch_size` as well and see.

Thanks,
Ahmed

**Comment by yelboudouri** [comment]
Hi, I started training the 'tiny' model for 30 epochs, and it ran as expected. As you can see in the figure below, both training and validation losses are decreasing. Are you sure you didn't change anything in the code? If so, try increasing the `early_stopping_patience` as suggested earlier.

![Screenshot from 2023-12-30 17-42-58](https://github.com/yelboudouri/EmoNeXt/assets/72700516/93c83fdb-b6a4-4868-9dfb-1fbe10efb8ef)

**Comment by anasserhussien** [comment]
Can you share your torch and torchvision versions?

**Comment by yelboudouri** [comment]
torch                            2.1.0+cu121
torchvision               0.16.0+cu121

**Comment by michiedevera** [comment]
Hi there!

May I know what worked to fix this problem? I'm having the same issue with training stopping at 12 epochs.

Thank you,
Michie

**Comment by Codenewwer** [comment]
> Hello,
> 
> Thanks for providing the codebase.
> 
> I was trying to replicate the experiment with the "tiny" model. I followed the steps provided for downloading the dataset and training the model. However, the training always stops at epoch 12 because the validation loss didn't improve. Any idea/guidance?
> 
> Best, Ahmed



> model. However, the training always stops at epoch 12 because the validation loss didn't improve. Any idea/guidance?
> 
> Best,

Hello,
I am trying to replicate the experiment using the "tiny" model. I downloaded the dataset and trained the model as per the steps provided. But at 178th epoch there is a case where the train loss and val loss is nan and the val acc and train val becomes 13% and it does not change for subsequent trainings，Any idea/guidance？
![train](https://github.com/yelboudouri/EmoNeXt/assets/99538869/b8e89c44-15bc-4031-8f06-9ffa52859b0c)
Best！
train

**Comment by Codenewwer** [comment]
> Hi there!
> 
> May I know what worked to fix this problem? I'm having the same issue with training stopping at 12 epochs.
> 
> Thank you, Michie

Hi！I met a new problem. The training accucacy always decreases from 38% to 13%. As shown in the figure below, How to solve it.Thank you. 
![train](https://github.com/yelboudouri/EmoNeXt/assets/99538869/1bd32da8-5615-41a2-8ca8-2eed2ec8bf0f)

**Comment by anasserhussien** [comment]
Hi @michiedevera and @Codenewwer,

For the `tiny` model I couldn't replicate the experiment and the accuracy always something below 40%.

However, I tried the `base` model with the default configurations except for the `batch_size=128` and I got a testing accuracy of `73.9760 %` . It's still lower than the acc obtained by the authors by `1 %` but at least it's a small gap.
 
I hope this helps you.

Best,
Ahmed

**Comment by Codenewwer** [comment]
> Hi @michiedevera and @Codenewwer,
> 
> For the `tiny` model I couldn't replicate the experiment and the accuracy always something below 40%.
> 
> However, I tried the `base` model with the default configurations except for the `batch_size=128` and I got a testing accuracy of `73.9760 %` . It's still lower than the acc obtained by the authors by `1 %` but at least it's a small gap.
> 
> I hope this helps you.
> 
> Best, Ahmed

Hi! Thank you very much for helping me. When you were reproducing the base model, did you just change the batch_size to 128? Did you leave everything else unchanged? Because when I reproduced it the way you did, the result was only 70.354%. Also xlarge model did you go to reproduce it? Because I also reproduced the xlarge model, but the result is only 71.273%.

**Comment by CugAgo** [comment]
> Hi @michiedevera and @Codenewwer,
> 
> For the `tiny` model I couldn't replicate the experiment and the accuracy always something below 40%.
> 
> However, I tried the `base` model with the default configurations except for the `batch_size=128` and I got a testing accuracy of `73.9760 %` . It's still lower than the acc obtained by the authors by `1 %` but at least it's a small gap.
> 
> I hope this helps you.
> 
> Best, Ahmed

May I ask how you resolved the issue of stopping the validation set after 12 epochs without improving its accuracy? Did you just modify the early_stopping_patience parameter? I tried Tiny and Base, but both stopped at the 12th epoch. 
Thank you

**Comment by Codenewwer** [comment]
> > Hi @michiedevera and @Codenewwer,
> > For the `tiny` model I couldn't replicate the experiment and the accuracy always something below 40%.
> > However, I tried the `base` model with the default configurations except for the `batch_size=128` and I got a testing accuracy of `73.9760 %` . It's still lower than the acc obtained by the authors by `1 %` but at least it's a small gap.
> > I hope this helps you.
> > Best, Ahmed
> 
> May I ask how you resolved the issue of stopping the validation set after 12 epochs without improving its accuracy? Did you just modify the early_stopping_patience parameter? I tried Tiny and Base, but both stopped at the 12th epoch. Thank you

Hi, you can try to retrain by lowering your learning rate

**Comment by Enspire-Academy** [comment]
> > Hi @michiedevera and @Codenewwer,
> > For the `tiny` model I couldn't replicate the experiment and the accuracy always something below 40%.
> > However, I tried the `base` model with the default configurations except for the `batch_size=128` and I got a testing accuracy of `73.9760 %` . It's still lower than the acc obtained by the authors by `1 %` but at least it's a small gap.
> > I hope this helps you.
> > Best, Ahmed
> 
> May I ask how you resolved the issue of stopping the validation set after 12 epochs without improving its accuracy? Did you just modify the early_stopping_patience parameter? I tried Tiny and Base, but both stopped at the 12th epoch. Thank you

I just adjusted the early_stopping_patience, it took me longer than 12 epochs for increase. I believe I eventually just turned off the early stopping and monitored it every now and then as it was training and manually stopped when it started overfitting.
